File: src/rag-module/embedding.ts
import { stableHash, tokenize } from "./utils";

export function textToDeterministicEmbedding(text: string, dim: number): number[] {
  const out = Array.from({ length: dim }, () => 0);
  const tokens = tokenize(text);
  if (!tokens.length) return out;

  for (const token of tokens) {
    const h = stableHash(token);
    const bucket = parseInt(h.slice(0, 8), 16) % dim;
    const sign = parseInt(h.slice(8, 16), 16) % 2 === 0 ? 1 : -1;
    out[bucket] += sign;
  }

  let norm = 0;
  for (let i = 0; i < out.length; i++) norm += out[i] * out[i];
  if (norm === 0) return out;
  const inv = 1 / Math.sqrt(norm);
  for (let i = 0; i < out.length; i++) out[i] *= inv;
  return out;
}

File: rag_python/embedding.py
from __future__ import annotations

import math
from typing import Protocol

from .utils import stable_hash, tokenize


class Embedder(Protocol):
    def encode(self, text: str) -> list[float]:
        ...


class DeterministicHashEmbedder:
    def __init__(self, dim: int = 512):
        self.dim = dim

    def encode(self, text: str) -> list[float]:
        out = [0.0] * self.dim
        tokens = tokenize(text)
        if not tokens:
            return out

        for token in tokens:
            h = stable_hash(token)
            bucket = int(h[:8], 16) % self.dim
            sign = 1.0 if int(h[8:16], 16) % 2 == 0 else -1.0
            out[bucket] += sign

        norm = math.sqrt(sum(x * x for x in out))
        if norm == 0:
            return out
        return [x / norm for x in out]
